test/macaroni/test/BasicActionsTest.java

```diff
@@ -0,0 +1,320 @@
+		assertTrue(setOutputSuccess);
+	}
+
+	@Test
+	void setPumpInput_InvalidPos() {
+		var pipe = new Pipe(wc);
+		var plumber = new Plumber(this.pipe);
+		boolean addSuccess = pump.addPipe(pipe);
+		boolean setInputSuccess = plumber.setInputPipe(pump, pipe);
+
+		assertTrue(addSuccess);
+		assertFalse(setInputSuccess);
+	}
+
+	@Test
+	void setPumpOutput_InvalidPos() {
+		var pipe = new Pipe(wc);
+		var plumber = new Plumber(this.pipe);
+		boolean addSuccess = pump.addPipe(pipe);
+		boolean setOutputSuccess = plumber.setOutputPipe(pump, pipe);
+
+		assertTrue(addSuccess);
+		assertFalse(setOutputSuccess);
+	}
+
+	@Test
+	void setPumpInput_AlreadyOutput() {
+		var pipe = new Pipe(wc);
+		var plumber = new Plumber(pump);
+		boolean addSuccess = pump.addPipe(pipe);
+		boolean setOutputSuccess = pump.setOutputPipe(pipe);
+		boolean setInputSuccess = plumber.setInputPipe(pump, pipe);
+
+		assertTrue(addSuccess);
+		assertTrue(setOutputSuccess);
+		assertFalse(setInputSuccess);
+	}
+
+	@Test
+	void setPumpOutput_AlreadyInput() {
+		var pipe = new Pipe(wc);
+		var plumber = new Plumber(pump);
+		boolean addSuccess = pump.addPipe(pipe);
+		boolean setInputSuccess = pump.setInputPipe(pipe);
+		boolean setOutputSuccess = plumber.setOutputPipe(pump, pipe);
+
+		assertTrue(addSuccess);
+		assertTrue(setInputSuccess);
+		assertFalse(setOutputSuccess);
+	}
+
+	@Test
+	void setPumpInput_AlreadyInput() {
+		var pipe = new Pipe(wc);
+		var plumber = new Plumber(pump);
+		boolean addSuccess = pump.addPipe(pipe);
+		boolean setInputSuccess1 = pump.setInputPipe(pipe);
+		boolean setInputSuccess2 = plumber.setInputPipe(pump, pipe);
+
+		assertTrue(addSuccess);
+		assertTrue(setInputSuccess1);
+		assertFalse(setInputSuccess2);
+	}
+
+	@Test
+	void setPumpOutput_AlreadyOutput() {
+		var pipe = new Pipe(wc);
+		var plumber = new Plumber(pump);
+		boolean addSuccess = pump.addPipe(pipe);
+		boolean setOutputSuccess1 = pump.setOutputPipe(pipe);
+		boolean setOutputSuccess2 = plumber.setOutputPipe(pump, pipe);
+
+		assertTrue(addSuccess);
+		assertTrue(setOutputSuccess1);
+		assertFalse(setOutputSuccess2);
+	}
+
+	@Test
+	void pickupPump() {
+		ModelObjectFactory.setCisternCreatePumpName("testPump");
+		var cistern = new Cistern(wc);
+		var plumber = new Plumber(cistern);
+
+		boolean pickupSuccess = plumber.pickUpPump(cistern);
+
+		assertTrue(pickupSuccess);
+		assertEquals(plumber.getHeldPumpCount(), 1);
+	}
+
+	@Test
+	void pickupPump_InvalidPos() {
+		ModelObjectFactory.setCisternCreatePumpName("testPump");
+		var cistern = new Cistern(wc);
+		var plumber = new Plumber(pipe);
+
+		boolean pickupSuccess = plumber.pickUpPump(cistern);
+
+		assertFalse(pickupSuccess);
+		assertEquals(plumber.getHeldPumpCount(), 0);
+	}
+
+	@Test
+	void placePump() {
+		var pipe = new Pipe(wc);
+		var pump1 = new Pump();
+		var pump2 = new Pump();
+		var pump3 = new Pump();
+		pump1.addPipe(pipe);
+		pump2.addPipe(pipe);
+		pipe.addEndpoint(pump1);
+		pipe.addEndpoint(pump2);
```

Comment:
addEndpoint is called automatically with the ActiveElement the addPipe function is called on, so there's no need to call it.